src/core/state.ml

```diff
@@ -18,28 +27,74 @@ let apply_tezos_operation t tezos_operation =
           let destination = key_hash in
           Ledger.deposit destination amount ticket ledger
         | Originated _ -> failwith "not implemented" in
-      { ledger } in
+      { t with ledger } in
   let { hash = _; payload } = tezos_operation in
   let { tezos_operation_hash = _; internal_operations } = payload in
   List.fold_left apply_internal_operation t internal_operations
+
 let apply_user_operation t user_operation =
   let open User_operation in
-  let { hash = _; sender; initial_operation } = user_operation in
+  let { hash; sender; initial_operation } = user_operation in
   match (Address.to_key_hash sender, initial_operation) with
   | Some sender, Transaction { destination; amount; ticket } ->
     let%ok ledger =
       Ledger.transfer ~sender ~destination amount ticket t.ledger in
-    Ok ({ ledger }, None)
+    Ok ({ t with ledger }, None)
   | None, Transaction _ -> Error `Transaction_sender_must_be_implicit_account
   | Some sender, Tezos_withdraw { owner; amount; ticket } ->
     let%ok ledger, handle =
       Ledger.withdraw ~sender ~destination:owner amount ticket t.ledger in
-    Ok ({ ledger }, Some (Receipt_tezos_withdraw handle))
+    Ok ({ t with ledger }, Some (Receipt_tezos_withdraw handle))
   | None, Tezos_withdraw _ -> Error `Withdraw_sender_must_be_implicit_account
+  | Some sender, Contract_origination { to_originate; ticket } ->
+    (* @TODO: deduct gas from account and check *)
+    let balance = Ledger.balance sender ticket t.ledger in
+    let origination_cost = 250 |> Amount.of_int in
+    let%assert () =
+      Amount.
+        ( `Origination_error
+            {
+              t with
+              ledger = Ledger.burn t.ledger ~sender ~ticket ~amount:balance;
+            },
+          let comparison_result = compare balance origination_cost in
+          comparison_result >= 0 ) in
+    let t =
+      {
+        t with
+        ledger = Ledger.burn t.ledger ~sender ~ticket ~amount:origination_cost;
+      } in
+    let initial_gas = Amount.to_int Amount.(balance - origination_cost) in
+    let burn_and_update t _ to_burn =
+      `Origination_error
+        {
+          t with
+          ledger =
+            Ledger.burn ~sender ~ticket ~amount:(to_burn |> Amount.of_int)
+              t.ledger;
+        } in
+    let%ok contract, to_burn =
+      Smart_contracts.Contract.Compile.compile_script ~originated_by:sender
+        ~gas:initial_gas ~on_error:(burn_and_update t) to_originate in
+    let address = hash |> BLAKE2B.to_raw_string |> BLAKE2B_20.hash in
+    let contract_storage =
+      Smart_contracts.Contract_storage.originate_contract t.contract_storage
+        ~address ~contract in
+    Ok
+      ( {
+          contract_storage;
+          ledger =
+            Ledger.burn t.ledger ~sender ~ticket ~amount:Amount.(of_int to_burn);
+        },
+        None )
+  | None, Contract_origination _ ->
+    Error `Contract_originator_must_be_implicit_account
```

Comment:
Change sender type from Address.t to Key_hash.t

Reply:
changed